palace/fem/bilinearform.hpp

```diff
@@ -0,0 +1,153 @@
+// SPDX-License-Identifier: Apache-2.0
+
+#ifndef PALACE_FEM_BILINEARFORM_HPP
+#define PALACE_FEM_BILINEARFORM_HPP
+
+#include <memory>
+#include <vector>
+#include <mfem.hpp>
+#include "fem/integrator.hpp"
+#include "fem/libceed/operator.hpp"
+
+namespace palace
+{
+
+//
+// This class implements bilinear and mixed bilinear forms based on integrators assembled
+// using the libCEED library. Assembly in the form of a partially assembled operator or
+// fully assembled sparse matrix is available.
+//
+class BilinearForm
+{
+protected:
+  // Domain and range finite element spaces.
+  const mfem::ParFiniteElementSpace &trial_fespace, &test_fespace;
+
+  // List of domain and boundary integrators making up the bilinear form.
+  std::vector<std::unique_ptr<BilinearFormIntegrator>> domain_integs, boundary_integs;
+
+  // Integration order for quadrature rules is calculated as p_trial + p_test + w + q_extra,
+  // where p_test and p_trial are the test and trial space basis function orders and w is
+  // the geometry order.
+  int q_extra_pk, q_extra_qk;
+
+public:
+  BilinearForm(const mfem::ParFiniteElementSpace &trial_fespace,
+               const mfem::ParFiniteElementSpace &test_fespace, int q_extra_pk,
+               int q_extra_qk)
+    : trial_fespace(trial_fespace), test_fespace(test_fespace), q_extra_pk(q_extra_pk),
+      q_extra_qk(q_extra_qk)
+  {
+  }
+  BilinearForm(const mfem::ParFiniteElementSpace &trial_fespace,
+               const mfem::ParFiniteElementSpace &test_fespace, int q_extra = 0)
+    : BilinearForm(trial_fespace, test_fespace, q_extra, q_extra)
+  {
+  }
+  BilinearForm(const mfem::ParFiniteElementSpace &fespace, int q_extra_pk, int q_extra_qk)
+    : BilinearForm(fespace, fespace, q_extra_pk, q_extra_qk)
+  {
+  }
+  BilinearForm(const mfem::ParFiniteElementSpace &fespace, int q_extra = 0)
+    : BilinearForm(fespace, fespace, q_extra, q_extra)
+  {
+  }
+
+  const auto &GetTrialSpace() const { return trial_fespace; }
+  const auto &GetTestSpace() const { return test_fespace; }
+
+  // MFEM's RT_FECollection actually returns order + 1 for GetOrder() for historical
+  // reasons.
+  auto GetMaxElementOrder() const
+  {
+    const auto &trial_fec = *trial_fespace.FEColl();
+    const auto &test_fec = *test_fespace.FEColl();
+    return std::max(
+        dynamic_cast<const mfem::RT_FECollection *>(&trial_fec) ? trial_fec.GetOrder() - 1
+                                                                : trial_fec.GetOrder(),
+        dynamic_cast<const mfem::RT_FECollection *>(&test_fec) ? test_fec.GetOrder() - 1
+                                                               : test_fec.GetOrder());
+  }
+
+  void AddDomainIntegrator(std::unique_ptr<BilinearFormIntegrator> &&bfi)
+  {
+    domain_integs.push_back(std::move(bfi));
+  }
+
+  void AddBoundaryIntegrator(std::unique_ptr<BilinearFormIntegrator> &&bfi)
+  {
+    boundary_integs.push_back(std::move(bfi));
+  }
```

Comment:
`make_unique` is just a bit of parameter pack trickery, so it's actually possible to bake the technique directly into this interface
```
  template <typename T, typename... Args>
  void AddDomainIntegrator(Args &&... args)
  {
    domain_integs.push_back(std::make_unique<T>(std::forward<Args>(args)...));
  }

  template <typename T, typename... Args>
  void AddBoundaryIntegrator(Args &&... args)
  {
    boundary_integs.push_back(std::make_unique<T>(std::forward<Args>(args)...));
  }
```
where the call sites are now shortened down to things like
```
a.AddDomainIntegrator<CurlCurlIntegrator>(*df);
k.AddDomainIntegrator<CurlCurlIntegrator>(muinv_func);
```
rather than needing to manually handle the pointers. This would apply to all of these `AddXXX` methods.

Reply:
Nice, I like that. Thanks!

Reply:
Resolved in [e830d8e](https://github.com/awslabs/palace/pull/97/commits/e830d8ea39d0c52eea8f94e00fc6ff7c4b9b3953).